Type checking

PeterJCLaw opened the pull request:
Adds (fairly relaxed) type checking.

PeterJCLaw committed:
Add initial types

This is _very_ work-in-progress and emits quite a lot of errors.

So far the main frustrations are:
 - lazy assignment of attributes (the code appears to actually be
   safe, though mypy can't see that and it's not clear how to
   encode this concept into the type signatures)
 - types at the boundaries of native code neccesitate some fudging
   (currently resolved by adding a types.pyi stub file)
 - numpy doesn't have types (though the https://github.com/machinalis/mypy-data
   looks a promising start); even if it did, we really want to be
   able to encode a 3x3 matrix as a different type than a 1x9 array,
   yet they're the same thing to numpy (both an ndarray)

PeterJCLaw committed:
Add types relating to the detector wrapper

PeterJCLaw committed:
Add missing import

PeterJCLaw committed:
Switch to class-based definitions of new types

PeterJCLaw committed:
Push the distance model into the base class

I'm not strictly sure the camera should have the distance model,
but since we rely on all cameras having one at the moment it
should be in the base class.

PeterJCLaw committed:
Add guard

PeterJCLaw committed:
Be much more lenient with our type checking

A number of our imports don't have stubs yet, so we need to
ignore them at least for now.

PeterJCLaw committed:
Ignore imports only used for typing

PeterJCLaw committed:
Push get_image_size into specific implementations

PeterJCLaw committed:
Explicitly annotate this

PeterJCLaw committed:
The distance model is optional

It turns out it was already optional, just mis-labelled.

RealOrangeOne commented on sb_vision/native/apriltag/detector.py:
```diff
@@ -1,14 +1,19 @@
 
 from ._apriltag import ffi, lib
 
+if False:
```
er....

PeterJCLaw replied:
This is the way to specify typing-only imports.

PeterJCLaw replied:
Python 3.6 has a better solution (a constant `TYPING` which type-checkers can treat as `True` but which the runtime treats as `False`), but we need Python 3.5 support.

PeterJCLaw replied:
It turns out that the constant I was thinking of ([`typing.TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING)) was actually introduced in Python 3.5.2, so we can _probably_ use it. I'll check which patch version of Python is on the Pis before switching over though.

PeterJCLaw replied:
It turns out that the Pis are running python 3.5.3, so this should be fine to use.

PeterJCLaw replied:
159b74c

RealOrangeOne commented on setup.cfg:
```diff
@@ -29,3 +29,15 @@ multi_line_output = 3
+strict_optional = True
+
+disallow_any_generics = True
+# disallow_untyped_calls = True
```
Can we remove this commented code?

PeterJCLaw replied:
980de983e

PeterJCLaw committed:
Remove commented config

I'd like to be able to reenable these options at some point, but
for now they're a distraction.

PeterJCLaw committed:
Explicitly tell flake8 to use setup.cfg for mypy

It seems ridiculous to me that we need to do this, but we do.

PeterJCLaw committed:
Tell mypy to ignore native import it doesn't understand

PeterJCLaw committed:
Prefer the TYPE_CHECKING constant over 'False'

PeterJCLaw commented:
@prophile @RealOrangeOne I think this is ready to go; the only thing I'm not sure about is whether we want to use `flake8-mypy` or not. My current thinking is that we should keep them separate so the lints are fast, but we should probably make a decision once and follow the same pattern for all our repos.

RealOrangeOne commented:
I'm certainly for _not_ using `flake8-mypy`. Most editors should support type checking ontop of linting, and it can always be installed locally if not. In tests it makes little difference, so I'm all for not using it.

RealOrangeOne commented on sb_vision/native/apriltag/detector.py:
```diff
@@ -1,14 +1,19 @@
 
 from ._apriltag import ffi, lib
 
+if TYPE_CHECKING:
```
It looks like there are a few places we're importing for only the sake of type definitions. Is this conditionally really necessary, or is it something overlooked in other locations

PeterJCLaw replied:
Does 8cf929567b86f73b334eb4e0900d80f4871196b5 help?

PeterJCLaw replied:
For clarity: in general I think we should do plain imports of types which are needed for type checking, however the constant is useful when either that would result in the import being circular or the types being imported aren't actually importable at runtime (as in this case).

PeterJCLaw committed:
Clarify why these conditionals are needed

PeterJCLaw committed:
Separate type checking from linting

This ensures that linting passes are nice and fast, which keeps
the short feedback cycle we want from it. We expect that developers
will configure their editor to run mypy directly in addition to
flake8 anyway, so this shouldn't have any impact on seeing mypy's
output.

PeterJCLaw commented:
I've separated `mypy` from `flake8` in 223dc07, using the same script naming pattern as is established in `runusb`.

PeterJCLaw committed:
Type check the right things..
